okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt

```diff
@@ -228,6 +228,9 @@ class FakeFileSystem(
 
     element.access(now = clock.now())
     val paths = elements.keys.filterTo(mutableListOf()) { it.parent == canonicalPath }
+    for (i in paths.indices) {
```

Comment:
Since workingDirectory is known to be absolute, it seems only worthwhile doing this if dir.isRelative == true.

Reply:
As a performance optimization? Or for different behavior?

Reply:
Performance optimisation, but the good kind, skipping work when you know it isn't needed :)

Working on ResourceFileSystem made me aware of how frequently we end up doing path comparisons in tight loops.  

Some thoughts, outside of this PR which is after all a Fake FileSystem.

1) They aren't particularly optimised since ultimately if length matches they become a front to back array comparison in ByteString. And given properties of a file system, the prefix is much more likely (not guaranteed) to match than suffix. What's the plan here? optimise each case like ZipFileSystem's fully populated index.
2) Should we use the cached hashCode of ByteString to avoid this work.  It might also minimise (poorly) string length oracle attacks.
3) We are getting into the territory of us implementing a lot of [accidentally quadratic](https://accidentallyquadratic.tumblr.com) library functions.

I still remember the terrible Java 1.0.2? string hashcode that skipped characters and was pathologically bad on structured filesystems (think a strict naming system for camera images).

Reply:
Implemented the performance optimization.

Using the cached hash code in equals is potentially tricky. We won’t always have a hash code for one or both strings, so short-circuiting on it only works if we have it for both of ’em. (I wonder how often we’ll have a hash code for both strings?!)

As for accidentally quadratic things . . . I think the biggest danger we face with performance surprises is in Buffers and internal segments. I’ve tried to make it so segments are 50% full or better whenever possible; this is partially to avoid quadratic hazards.

Reply:
Yeah, most of this was me planting the thought. Since it's been bugging me. I've seen weird cases when Intellij freezes (10-15 seconds) turned out to be assumptions made in jetbrains library code that broke on FB or Google size codebase. 

On a large directory, with a very long prefix...

Nothing urgent.